### PR TITLE
[CUDA][cuBLAS] Add fp16 accumulate option to cuBLAS/cuBLASLt

### PR DESCRIPTION
**Pull request by eqy**
Test for `cublasGemmEx` added, still need to figure out the best way to exercise the other APIs...

cc @ptrblck @msaroufim @csarofeen @xwang233

### CODE REVIEW

File: docs/source/notes/cuda.rst [hunk 1]
```diff
@@ -148,6 +148,8 @@ For more information about TF32, see:
 Reduced Precision Reduction in FP16 GEMMs
 -----------------------------------------
 
+(Distinct from full FP16 accumulation, see :ref:`Full FP16 accumulation<fp16accumulation>`)
```
Comment (Chillee) [inline docs/source/notes/cuda.rst:151]: Might be worth elaborating a bit more on what exactly this means.

File: docs/source/notes/cuda.rst [hunk 2]
```diff
@@ -206,6 +208,24 @@ To toggle the reduced precision reduction flags in C++, one can do
...
+
+Full FP16 Accmumulation in FP16 GEMMs
+-------------------------------------
+
+Certain GPUs have increased performance when doing _all_ FP16 GEMM accumulation
+in FP16, at the cost of numerical precision. This behavior can be enabled via:
+
+.. code:: python
+
+  torch.backends.cuda.matmul.use_fp16_accumulation = True
```
Comment (Chillee) [inline docs/source/notes/cuda.rst:221]: I think it would be good to align these two flags a bit more. Perhaps just `allow_fp16_reduced_precision_accumulation`?

**Review by Chillee — APPROVED** [review]

File: aten/src/ATen/Context.cpp [hunk 3]
```diff
@@ -411,6 +410,14 @@ void Context::setAllowBF16ReductionCuBLAS(bool b) {
   allow_bf16_reduction_cublas = b;
 }
 
+bool Context::useFP16AccumulationCuBLAS() const {
+  return use_fp16_accumulation_cublas;
```
Comment (liangan1) [inline aten/src/ATen/Context.cpp:414]: what is the difference between useFP16AccumulationCuBLAS and allowFP16ReductionCuBLAS?

Comment (eqy) [reply aten/src/ATen/Context.cpp:414]: the reduction flag is not doing all accumulations in FP16 (less likely to have numerical issues like overflow) where as the accumulation flag will do all accumulations in FP16 (more likely to have overflow)

File: docs/source/notes/cuda.rst (hunk 2 above)
Comment (eqy) [reply docs/source/notes/cuda.rst:221]: yeah I think that makes sense, was initially waffling about this as technically the reduction flag is leaving it up to heuristics whether or not reduced precision reductions were used, but fp16 accumulation is more heavy-handed in that it requires scales to be in fp16 as well (I was thinking that implies that it's _always_ going to use fp16 accum)

**Comment by Chillee** [comment]
Oh also, is there bf16 + bf16 accumulation?

**Comment by eqy** [comment]
I don't see that in the support table
image

**Comment by Birch-san** [comment]
the loss of accuracy (versus fp32) can be mitigated (whilst still retaining the favourable performance) by using split-k matmul.

fluffy measured the accuracy and perf for various values of k in a triton implementation:  
https://github.com/sekstini/gpupoor

Zippy offers a cuBLAS implementation here:  
https://github.com/aredden/torch-cublas-hgemm

**Comment by Birch-san** [comment]
3090/4090 users would also likely benefit from fp16 accumulation inside attention and convolution operations. with this, perhaps the 4090 could approximate the speed of an A100.

**Comment by FurkanGozukara** [comment]
Awesome I hope it comes and works flawlessly

**Comment by FurkanGozukara** [comment]
so once this is merged how we gonna use / enable this?

**Comment by eqy** [comment]
> Referen

see the built docs: https://docs-preview.pytorch.org/pytorch/pytorch/144441/notes/cuda.html#full-fp16-accmumulation-in-fp16-gemms

ultimately it's on the user to validate that their training/inference recipes work with the reduced dynamic range

**Comment by kit1980** [comment]
There are many Meta-internal builds failed for unused variables.
@Chillee maybe you can help shepherd this? D68059314

**Comment by eqy** [comment]
@kit1980 would it be possible to share the line numbers that the warning was emitted on?
might be a quick fix

**Comment by Chillee** [comment]
@eqy All in CUDABlas:

	1.	550
	2.	551
	3.	552
	4.	553
	5.	554
	6.	899
	7.	900
	8.	901
	9.	902
	10.	905

**Comment by kit1980** [comment]
@eqy All error are "unused variables".
But the errors are not when building PyTorch itself, but when building projects that use PyTorch (hundreds of failed build targets)

```
550 |   at::Half halpha;
551 |   at::Half hbeta;
552 |   void * alpha_ptr = &falpha;
553 |   void * beta_ptr = &fbeta;
554 |   auto compute_type = HIP_R_32F;
899 |   at::Half halpha;
900 |   at::Half hbeta;
901 |   void * alpha_ptr = &falpha;
902 |   void * beta_ptr = &fbeta;
905 |   auto compute_type = HIP_R_32F;
```

**Comment by Chillee** [comment]
@kit1980 They're all during hipification I think.

**Comment by kit1980** [comment]
> They're all during hipification I think.

Correct.

**Comment by GrigoryEvko** [comment]
> the loss of accuracy (versus fp32) can be mitigated (whilst still retaining the favourable performance) by using split-k matmul.
> 
> fluffy measured the accuracy and perf for various values of k in a triton implementation: https://github.com/sekstini/gpupoor
> 
> Zippy offers a cuBLAS implementation here: https://github.com/aredden/torch-cublas-hgemm

Super hyped to see this implemented in pytorch!
image
image
@kit1980 @Chillee @eqy @Skylion007

**Comment by kit1980** [comment]
@eqy @Chillee I see the same errors as before internally

**Comment by eqy** [comment]
Ah my bad, I think I dropped my last commit...

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m "unused variables again :(" -c ghfirst


looks like the same issue again when USE_ROCM is not defined.
```
Stderr:
.../HIPBlas.cpp:554:10: error: unused variable 'alpha_ptr' [-Werror,-Wunused-variable]
  554 |   void * alpha_ptr = &falpha;
      |          ^~~~~~~~~
.../HIPBlas.cpp:555:10: error: unused variable 'beta_ptr' [-Werror,-Wunused-variable]
  555 |   void * beta_ptr = &fbeta;
      |          ^~~~~~~~
.../HIPBlas.cpp:556:8: error: unused variable 'compute_type' [-Werror,-Wunused-variable]
  556 |   auto compute_type = HIP_R_32F;
      |        ^~~~~~~~~~~~
.../HIPBlas.cpp:905:10: error: unused variable 'alpha_ptr' [-Werror,-Wunused-variable]
  905 |   void * alpha_ptr = &falpha;
      |          ^~~~~~~~~
.../HIPBlas.cpp:906:10: error: unused variable 'beta_ptr' [-Werror,-Wunused-variable]
  906 |   void * beta_ptr = &fbeta;
      |          ^~~~~~~~
.../HIPBlas.cpp:909:8: error: unused variable 'compute_type' [-Werror,-Wunused-variable]
  909 |   auto compute_type = HIP_R_32F;
      |        ^~~~~~~~~~~~
6 errors generated.
```

**Comment by ZainRizvi** [comment]
@pytorchbot revert -c ghfirst -m "Sorry but this is failing internally. @Chillee can you please help change get remerged? See  D68720562"

Error message:
```
Stderr:
...hip/HIPBlas.cpp:556:8: error: unused variable 'compute_type' [-Werror,-Wunused-variable]
  556 |   auto compute_type = HIP_R_32F;
      |        ^~~~~~~~~~~~
.../HIPBlas.cpp:909:8: error: unused variable 'compute_type' [-Werror,-Wunused-variable]
  909 |   auto compute_type = HIP_R_32F;
      |        ^~~~~~~~~~~~
2 errors generated.
```

**Review by malfet — APPROVED** [review]
Amount of glue code is a bit sad, would be nice have more code reuse...

**Comment by ZainRizvi** [comment]
@pytorchbot revert -c ghfirst -m "Sorry but this still fails internally with the same error.  @Chillee or @malfet, can you please help the change get tested? (See D68783351)"

Instructions for testing the changes internally: https://fburl.com/fixing-ghfirst-reverts

```
Stderr:
...hip/HIPBlas.cpp:556:8: error: unused variable 'compute_type' [-Werror,-Wunused-variable]
  556 |   auto compute_type = HIP_R_32F;
      |        ^~~~~~~~~~~~
.../HIPBlas.cpp:909:8: error: unused variable 'compute_type' [-Werror,-Wunused-variable]
  909 |   auto compute_type = HIP_R_32F;
      |        ^~~~~~~~~~~~
2 errors generated.
```
